src/main/java/com/google/devtools/build/lib/exec/SpawnInputExpander.java

```diff
@@ -251,7 +258,14 @@ private static void addInputs(
         ActionInputHelper.expandArtifacts(
             inputFiles, artifactExpander, /* keepEmptyTreeArtifacts= */ true);
     for (ActionInput input : inputs) {
-      addMapping(inputMap, input.getExecPath(), input, baseDirectory);
+      if (input instanceof TreeFileArtifact) {
+        addMapping(inputMap,
+            pathMapper.map(((TreeFileArtifact) input).getParent().getExecPath())
+                .getRelative(((TreeFileArtifact) input).getParentRelativePath()), input,
```

Comment:
> Side note: I'd strongly prefer PathMapper.map to have a ActionInput->PathFragment signature (or perhaps (ActionInput, PathFragment)->PathFragment to handle the runfiles case) instead of PathFragment->PathFragment: it's a lot less error-prone to enforce restrictions like this one in the mapper implementation instead of risking a caller getting it wrong. But I suspect there's a good reason we decided not to.

I fully agree. Most of the preparatory work for the current PRs was to reduce the variety of functions on the (previously multiple) path mapping related interfaces down to a more manageable number. My favorite approach would be to get this into an end-to-end functional state within Bazel with good test coverage first and then refactor it into an API that is harder to misuse. Doing the second step now would risk missing use cases and then having to expand the API again.

Reply:
Sounds good, let's revisit it later once we're confident that we got everything to work.